### PR TITLE
br: Fix stuck when meeting error

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/33200

Problem Summary:

### What is changed and how it works?
This make the worker of `TiKVSender` select over the context created by the error group(instead of its parent context), so they can get aware of the error happened and exit properly.

This also make an interface `TiKVRestorer`, which is the backend for `TiKVSender`, these can make us easier to writing unit tests.

### Check List

Tests 

- [x] Unit test

### Release note

```release-note
Fix a bug that caused BR get stuck when restore meets some unrecoverable error.
```

### CODE REVIEW

File: br/pkg/restore/split_test.go [hunk 1]
```diff
@@ -560,3 +565,123 @@ func TestRegionConsistency(t *testing.T) {
...
+	dctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	sink := assertErrorEmitInTime(dctx, t)
+	sender.PutSink(sink)
+	for _, r := range ranges {
+		sender.RestoreBatch(r)
+	}
+	sink.Wait()
+	sender.Close()
+	require.GreaterOrEqual(t, len(r.splitRanges), 2)
```
Comment (3pointer) [inline br/pkg/restore/split_test.go:685]: why 2 not 1 🤔

Comment (YuJuncen) [reply br/pkg/restore/split_test.go:685]: There are 2 ranges in a batch.

**Comment by YuJuncen** [comment]
[](https://ci.pingcap.net/blue/organizations/jenkins/tidb_ghpr_unit_test/detail/tidb_ghpr_unit_test/41482/#step-73-log-485)[2022-03-17T07:58:10.031Z] [FAIL]  planner/core TestCompareMultiFieldsInSubquery
[2022-03-17T07:58:10.031Z] err=signal: segmentation fault (core dumped)

🤔 /run-unit-test

File: br/pkg/restore/split_test.go [hunk 2]
```diff
@@ -560,3 +565,123 @@ func TestRegionConsistency(t *testing.T) {
...
+		err := errors.Annotatef(berrors.ErrRestoreSplitFailed,
+			"the key space takes many efforts and finally get together, how dare you split them again... :<")
+		log.Error("error happens :3", logutil.ShortError(err))
+		return err
+	}
+	return nil
+}
+
+func (f *fakeRestorer) RestoreFiles(ctx context.Context, files []*backuppb.File, rewriteRules *restore.RewriteRules, updateCh glue.Progress) error {
+	if ctx.Err() != nil {
```
Comment (3pointer) [inline br/pkg/restore/split_test.go:590]: 🤔 do we have similar code in real RestoreFiles?

File: br/pkg/restore/split_test.go (hunk 1 above)
Comment (3pointer) [reply br/pkg/restore/split_test.go:685]: i see

**Review by 3pointer — COMMENTED** [review]
rest LGTM

File: br/pkg/restore/split_test.go (hunk 2 above)
Comment (YuJuncen) [reply br/pkg/restore/split_test.go:590]: Probably, but hidden in some internal call of the clients.

Comment (YuJuncen) [reply br/pkg/restore/split_test.go:590]: For example, here:
image

**Review by 3pointer — APPROVED** [review]

**Review by joccau — APPROVED** [review]
LGTM

**Pull request merged**